### ISSUE TITLE
Samba based backups erroneously fail with samba >= 4.1.6

### ISSUE THREAD

**Issue by cramblda** [backuppc/backuppc#14]
Samba version 4.1.6 or greater changed the way verbose output is handled. Per file backups status lines are no longer being generated by samba client. BackupPC expects these per file lines to exist in the output, or BackupPC reports that no files were dumped for the share and marks good backups as _failed_.

**Workaround**: You can workaround the issue with this patch: https://bugzilla.redhat.com/show_bug.cgi?id=1294761
and Using the BackupPC config option (Under "Schedule"):

```
    $Conf{BackupZeroFilesIsFatal} = 0;
```

Edit: fixed typo in config example

**Comment by xtaran** [comment]
That patch works fine for me against BackupPC 3.3.0 in Debian Stable.

**Comment by xtaran** [comment]
@cramblda: https://bugzilla.redhat.com/show_bug.cgi?id=1294761#c3 actually says the opposite:

```
$Conf{BackupZeroFilesIsFatal} = 0;
```

And disabling that check since it can't be checked anymore makes more sense to me, too.

**Comment by cramblda** [comment]
Yes you're right, that was a typo - I have updated my post. I copied and pasted that line from on older backuppc mail list thread I was working on (pre-dating the updates to the bug report I posted) and wasn't very careful obviously when creating this issue.

I'm glad this is working for you. Unfortunately, this is still just a workaround. We probably still need to determine how to get the verbose file listing back from samba if possible.

**Comment by ldrolez** [comment]
I have notices that in the new code of smbclient there's a possibility to get the transfered file name:

line 903 of clitar.c in tar_get_file():
`DBG(5, ("+++ %s\n", full_dos_path));`

So by running smbclient in verbose level 5 (-d 5), and adding a regexp in backuppc, we can avoid the 'BackupZeroFilesIsFatal' hack.
But maybe, it's easier to patch smbclient to get the old functionality back?

**Comment by ldrolez** [comment]
In my latest Debian upload I've added also the following patch, to take into account the '+++' lines of the debug 5 level.

```
--- backuppc-3.3.0.orig/lib/BackupPC/Xfer/Smb.pm
+++ backuppc-3.3.0/lib/BackupPC/Xfer/Smb.pm
@@ -217,7 +217,7 @@
         # This section is highly dependent on the version of smbclient.
         # If you upgrade Samba, make sure that these regexp are still valid.
         #
-        if ( /^\s*(-?\d+) \(\s*\d+[.,]\d kb\/s\) (.*)$/ ) {
+        if ( /^\s*(-?\d+) \(\s*\d+[.,]\d kb\/s\) (.*)$/ || /^tar:(\d+)\s+\+\+\+ (.*)$/ ) {
             my $sambaFileSize = $1;
             my $pcFileName    = $2;
             (my $fileName = $pcFileName) =~ s/\\/\//g;
```

But If you try smbclient, in debug level 5 you'll see that it's very verbose. I think we should write a smbclient patch instead...

**Comment by cramblda** [comment]
We could write a patch for the samba client, but we may want to check with the samba folks first and see what their goal was for this behavior modification. It may be there is another way to get the info we desire. We would want to make sure they would be interested in committing such a patch.

**Comment by ldrolez** [comment]
It seems that smbclient was completly rewritten, and some code was lost...
I think that the patch could be as simply adding the following line:

```
DBG(1, ("%12.0f (0.0 kb/s) %s\n", (double)finfo->size, full_dos_path));
```

**Comment by cramblda** [comment]
That does seem like a very simple patch. You could submit and see what the samba team says.

**Comment by gsmitheidw** [comment]
Personally I just migrated Windows host to use rsync. For me that seems easier until this has all been resolved via updated distro package than editing lines of Perl or rolling back samba and fixing various dependencies that become broken. 

One line of C code it may be, but that doesn't include all the time and hassle of dev libraries and compiling etc. 

I'm a Sysadmin not a developer, if I'm down at C code level something is terribly wrong with the software (although in this case it looks more like Samba than BackupPc is the cause)
But having said that, it is still interesting to see what went wrong. 




_____________________________________
Graham Smith
Technical Officer
Department of Computing
Institute of Technology Tallaght
Tallaght
Dublin 24
Ireland
________________________________________

**Comment by cramblda** [comment]
@gsmitheidw  Our plan is to get Samba updated, or update BackupPC to work with samba's recent change. We not planning on requiring anyone to permanently patch samba client or BackupPC on their  system.

**Comment by ldrolez** [comment]
For failing restores, it's the same: newer smbclients output nothing, except in devel level 5, where you can have see something like:

`tar:631  tar_process done, err = 0`

**Comment by ldrolez** [comment]
Debian users can manually install the patched package here : https://packages.debian.org/sid/backuppc
The install should work on Jessie too.

**Comment by gsmitheidw** [comment]
Just for anybody who has migrated from samba to rsyncd for Windows hosts temporarily, beware you may need to amend your exceptions. My current run is ignoring those and seems to be backing up _everything_ from the remote system. Also I noticed it was a long time before I saw some data being transferred - having read older posts etc it seems that a generation of filelisting happens first and that can be lengthy if like me you have a vast amount of small files. I could see RAM filling up a bit on the backuppc side. Then the transfer log for the server has been a bit slow to update - seems to be in batches but I can see data happening using iostat (from sysstat) - 26 IOPS (tps)

**Comment by gsmitheidw** [comment]
Just to say the backuppc patch is now available via the standard debian apt sources.
I've migrated back to smb for now. I have noticed that at the start of a backup task, the transfer log is taking a particularly long time to be created. My normal smb backup run is quite long ~3 days and i'm about 14 hrs into the current job and still no XferLOG.z appearing. I can see that things seems to be "happening" using lsof, but very vague. I assume something is being collated/buffered in memory or somewhere obscure and hasn't yet been committed to disk. Is there another way to see current progress other than ```/usr/share/backuppc/bin/BackupPC_zcat /var/lib/backuppc/pc/hostname/XferLOG.z | tail``` ?

**Comment by gsmitheidw** [comment]
Ok I found I needed to append a star to the end of my backup only files paths. That was causing a delay and ultimately a fail. I was also seeing errornous duplicate subfolders in the error log for that client.

**Comment by nvolpini** [comment]
Here a patch that makes "smb" backups and restores work again. Generated on a Ubuntu 16.04 fresh install with backuppc 3.3.1-2ubuntu3 and samba 2:4.3.9+dfsg-0ubuntu0.16.04.2

- Patch the file /usr/share/backuppc/lib/BackupPC/Xfer/Smb.pm with the attached patch
- Review backuppc option to be $Conf{BackupZeroFilesIsFatal} = '0';
- Review backuppc option $Conf{SmbClientRestoreCmd} = '$smbClientPath \\\\$host\\$shareName $I_option -U $userName -E -d 5 -c tarmode\\ full -Tx -';
The catch is to change the "-d 1" to "-d 5"

[Smb.pm.patch.txt](https://github.com/backuppc/backuppc/files/378419/Smb.pm.patch.txt)

**Comment by ldrolez** [comment]
Yes this patch is already available in Debian unstable and testing.

**Comment by ldrolez** [comment]
https://github.com/ldrolez/backuppc-debian/commit/3fc591ad4a2767e616c848017588e1015cda9a20

**Comment by maksyms** [comment]
Thanks for this. I also produced a different patch for this issue, logged under #21 and a pull request #22, which doesn't require running smbclient at debug level 5 (quite noisy) and it does not enforce specific configuration setting like `$Conf{BackupZeroFilesIsFatal} = '0';` Have a look which one you prefer.

Arguably, I did miss patching View.pm, but that is not essential if debug level for smbclient still stays at 1.
